api/middleware/sourceThrottler.go

```diff
@@ -0,0 +1,65 @@
+package middleware
+
+import (
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/gin-gonic/gin"
+)
+
+// globalThrottler is a middleware limiter used to limit total number of requests originating from the same source
+type sourceThrottler struct {
+	mutRequests    sync.Mutex
+	sourceRequests map[string]uint32
+	maxNumRequests uint32
+}
+
+// NewSourceThrottler creates a new instance of a sourceThrottler
+func NewSourceThrottler(maxNumRequests uint32) (*sourceThrottler, error) {
+	if maxNumRequests == 0 {
+		return nil, ErrInvalidMaxNumRequests
+	}
+
+	return &sourceThrottler{
+		mutRequests:    sync.Mutex{},
+		sourceRequests: make(map[string]uint32),
+		maxNumRequests: maxNumRequests,
+	}, nil
+}
+
+// Limit returns the handler func used by the gin server to limit requests originating from the same source
+func (st *sourceThrottler) Limit() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		remoteAddr, _, err := net.SplitHostPort(c.Request.RemoteAddr)
+		if err != nil {
+			_ = c.AbortWithError(http.StatusInternalServerError, err)
+			return
+		}
+
+		st.mutRequests.Lock()
+		requests := st.sourceRequests[remoteAddr]
+		isQuotaReached := requests >= st.maxNumRequests
+		st.sourceRequests[remoteAddr]++
+		st.mutRequests.Unlock()
+
+		if isQuotaReached {
+			c.AbortWithStatus(http.StatusTooManyRequests)
+			return
+		}
+
+		c.Next()
+	}
+}
+
+// Reset resets all accumulated counters
+func (st *sourceThrottler) Reset() {
```

Comment:
If we are using this approach we should know there is an edge - we can actually hit with double requests/interval than set up in config - we just have to catch the reseting point between request batches.
Isn't there a middleware package from a third party that would have bees useful?

Reply:
Agree, this will still plague the p2p limiters also as their implementations follow the same principles. This burst can be achieved once on two rounds thus. Tried to keep the implementation as lightweight as it could even with the above mentioned problem. If we spot an attack, we can lower the configurable thresholds. 

I have searched for some already implemented middleware limiters for the gin web server but found simple solutions like the one implemented here. Decided to go with our own variant to match the implementation code with the rest of the code base.